cmd/juju/commands/switch.go

```diff
@@ -74,10 +91,104 @@ func (c *switchCommand) Info() *cmd.Info {
 	})
 }
 
+// SetFlags implements Command.SetFlags.
+func (c *switchCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.CommandBase.SetFlags(f)
+	f.StringVar(&c.modelName, "m", "", "Model to operate in. Accepts [<controller name>:]<model name>")
+	f.StringVar(&c.modelName, "model", "", "")
+	f.StringVar(&c.controllerName, "c", "", "Controller to operate in")
+	f.StringVar(&c.controllerName, "controller", "", "")
+}
+
 func (c *switchCommand) Init(args []string) error {
-	var err error
-	c.Target, err = cmd.ZeroOrOneArgs(args)
-	return err
+	if c.modelName != "" && c.controllerName != "" {
+		return errors.Trace(fmt.Errorf("cannot specify both a --model and --controller"))
+	}
+
+	if c.controllerName != "" {
+		err := cmd.CheckEmpty(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else if c.modelName != "" {
+		err := cmd.CheckEmpty(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// This means we can support arguments like `juju switch -m mycontroller:mymodel`
+		c.parseModelName(c.modelName)
+	} else {
+		target, err := cmd.ZeroOrOneArgs(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// if the target does not contain a ":", it is an ambiguous target so
+		// we cannot parse it as a model name
+		if strings.Contains(target, ":") {
+			c.parseModelName(target)
+		} else {
+			c.controllerOrModelName = target
+		}
+	}
+
+	// expand "-" syntactic sugar where it's found
+	if c.controllerName == "-" {
+		previous, _, err := c.Store.PreviousController()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.controllerName = previous
+	}
+
+	if c.modelName == "-" {
+		controller, err := c.Store.CurrentController()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		previous, err := c.Store.PreviousModel(controller)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.modelName = previous
+	}
+
+	if c.controllerOrModelName == "-" {
+		c.controllerOrModelName = ""
+
+		previousController, changedController, err := c.Store.PreviousController()
+		if err != nil && !errors.Is(err, errors.NotFound) {
+			return errors.Trace(err)
+		}
+
+		// if the last switch was intra-controller (i.e. changedController is true), we need
+		// to figure out from which model and switch back there. Otherwise (i.e. inter-controller),
+		// it is sufficient to switch to just the previous controller.
+		if changedController {
+			c.controllerName = previousController
+		} else {
+			c.controllerName, err = c.Store.CurrentController()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			previousModel, err := c.Store.PreviousModel(c.controllerName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.modelName = previousModel
+		}
+	}
+
+	return nil
+}
```

Comment:
`errors.Trace` should (almost) all be replaced with `errors.Annotate` or `errors.Errorf`

Reply:
I will add it on Run function. Seems that annotate in Init are discarded anyway (even with --debug)